In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

# Time and Pivot Tables: Tasks

# Task 1: The following transaction data is available for the period from 01.12.2010 to 12.09.2011:

<font size="5">InvoiceNo — transaction number</font>  
<font size="5">StockCode — product code</font>  
<font size="5">Description — product description</font>    
<font size="5">Quantity — number of units added to the order</font>    
<font size="5">InvoiceDate — transaction date</font>      
<font size="5">UnitPrice — price per product unit</font>
<font size="5">CustomerID — client ID</font>  
<font size="5">Country — country of residence of the client</font>  

## In this task, the data read path is stored in the `path_to_file` variable.
## Import pandas and read the data with ISO-8859-1 encoding. Save the resulting DataFrame as `retail`, and the column names in the variable `retail_columns`.

In [3]:
# Here type of compression is zip
retail = pd.read_csv('D:/курс_архив/my_files/lesson_5/5_data.csv.zip',encoding='ISO-8859-1', compression='zip')

In [4]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

# Task 2: Check if there are any duplicate entries in the data, and specify their count as the answer. If duplicates exist, remove them from `retail`.

In [5]:
# Check the number of rows in the table  
retail.shape

(541909, 8)

In [6]:
# Remove duplicates from the table and save to a variable
retail_no_duplicates = retail.drop_duplicates()

In [44]:
retail_no_duplicates

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[536641 rows x 8 columns]

In [45]:
# Check the total number of rows in the table again after removing duplicates. 
retail_no_duplicates.shape

(536641, 8)

In [46]:
# Now subtract the row count of the de-duplicated table from the original table.  
541909-536641 

5268

# Answer: 5268

# Task 3: The data contains records of both successful transactions and cancellations. If a user canceled an order, a "C" (canceled) is added to the beginning of the transaction number (`InvoiceNo`).  
<font size="5">How many transactions were canceled by users? Each row counts as a separate transaction.</font>  

In [7]:
# First, check the data type of the "InvoiceNo" column
retail_no_duplicates.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

In [8]:
# Filter records where "InvoiceNo" starts with "C" and check the number of entries in the table.    
retail_no_duplicates[retail_no_duplicates['InvoiceNo'].str.startswith('C')]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[9251 rows x 8 columns]

#  Answer: 9251

# Task 4: Filter the data in `retail` to include only orders where `Quantity > 0`. Specify the remaining row count as the answer.

In [9]:
# Filter the table using the `query` method on the `Quantity` column and count the records by `InvoiceNo`   
retail_no_duplicates.query('Quantity > 0') \
    .agg({'InvoiceNo': 'count'})

InvoiceNo    526054
dtype: int64

# Answer: 526054

# Task 5: Calculate the number of orders for each user (`CustomerID`) from Germany. Keep only those who made more than N transactions (`InvoiceNo`), where N is the 80th percentile. Save the resulting user IDs in `germany_top` (just the IDs, not the entire DataFrame).

<font size="5">The filtered data is stored in `retail`. The order identifier is `InvoiceNo`. Each order can have more than one line in the data.</font>


In [10]:
# Filter data by country "Germany" to keep only orders from Germany.
# Group data by 'CustomerID' to count the unique orders for each user.
# Calculate the count of unique orders ('InvoiceNo') for each user.
# Rename the 'InvoiceNo' column to 'Orders_count' for readability.
# Keep only those users who made more orders than the 80th percentile in order count.  
# Select only the user IDs (`CustomerID`) from the filtered data..
germany_top = retail_no_duplicates.query('Country == "Germany"') \
    .groupby('CustomerID', as_index=False) \
    .agg({'InvoiceNo': 'nunique'}) \
    .rename(columns={'InvoiceNo': 'Orders_count'}) \
    .query('Orders_count > Orders_count.quantile(0.8)').CustomerID

# Answer:

In [51]:
germany_top

3     12471.0
4     12472.0
6     12474.0
8     12476.0
12    12481.0
16    12500.0
37    12569.0
44    12600.0
52    12619.0
53    12621.0
57    12626.0
65    12647.0
70    12662.0
78    12705.0
79    12708.0
80    12709.0
82    12712.0
84    12720.0
Name: CustomerID, dtype: float64

# Task 6: Using the object with user IDs (`germany_top`) from the previous step, filter the data to retain records only for the relevant users. Save the resulting DataFrame as `top_retail_germany`.

<font size="5">DataFrame is saved in `retail`, the necessary user IDs are in `germany_top`.</font>


# Answer:

In [54]:
top_retail_germany = retail.query('CustomerID in @germany_top')
top_retail_germany

InvoiceNo StockCode                          Description  Quantity  \
1109      536527     22809              SET OF 6 T-LIGHTS SANTA         6   
1110      536527     84347  ROTATING SILVER ANGELS T-LIGHT HLDR         6   
1111      536527     84945   MULTI COLOUR SILVER T-LIGHT HOLDER        12   
1112      536527     22242        5 HOOK HANGER MAGIC TOADSTOOL        12   
1113      536527     22244           3 HOOK HANGER MAGIC GARDEN        12   
...          ...       ...                                  ...       ...   
541726    581570     22139     RETROSPOT TEA SET CERAMIC 11 PC          3   
541727    581570     23077                  DOUGHNUT LIP GLOSS         20   
541728    581570     20750             RED RETROSPOT MINI CASES         2   
541729    581570     22505            MEMO BOARD COTTAGE DESIGN         4   
541730    581570      POST                              POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID  Country  
1109    12/1/2010 13:04       2.95     12662.0  Germany  
1110    12/1/2010 13:04       2.55     12662.0  Germany  
1111    12/1/2010 13:04       0.85     12662.0  Germany  
1112    12/1/2010 13:04       1.65     12662.0  Germany  
1113    12/1/2010 13:04       1.95     12662.0  Germany  
...                 ...        ...         ...      ...  
541726  12/9/2011 11:59       4.95     12662.0  Germany  
541727  12/9/2011 11:59       1.25     12662.0  Germany  
541728  12/9/2011 11:59       7.95     12662.0  Germany  
541729  12/9/2011 11:59       4.95     12662.0  Germany  
541730  12/9/2011 11:59      18.00     12662.0  Germany  

[4698 rows x 8 columns]

# Task 7: Group `top_retail_germany` by product code (`StockCode`). Which product was added to the cart most frequently, excluding POST?

In [38]:
# Filter data, excluding the product with code "POST"
# Group the filtered data by product code ('StockCode') 
# Count the number of transactions for each product (by the number of unique 'InvoiceNo')  
# Sort products by transaction count in descending order
top_retail_germany.query('StockCode != "POST"') \
    .groupby('StockCode', as_index=False) \
    .agg({'InvoiceNo': 'count'}) \
    .sort_values('InvoiceNo', ascending =False)

StockCode  InvoiceNo
409     22326         63
452     22423         55
411     22328         45
474     22467         37
46      20719         34
..        ...        ...
24      16238          1
22      16236          1
20      16218          1
18     16168M          1
16     16161P          1

[1167 rows x 2 columns]

# Answer: StockCode: 22326

# Task 8: Return to analyzing the complete dataset `retail`. Create a `Revenue` column with the purchase amount using `Quantity` and `UnitPrice`.

In [11]:
retail['Revenue'] = retail.Quantity * retail.UnitPrice
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  Revenue  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom    15.30  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom    22.00  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
...                 ...        ...         ...             ...      ...  
541904  12/9/2011 12:50       0.85     12680.0          France    10.20  
541905  12/9/2011 12:50       2.10     12680.0          France    12.60  
541906  12/9/2011 12:50       4.15     12680.0          France    16.60  
541907  12/9/2011 12:50       4.15     12680.0          France    16.60  
541908  12/9/2011 12:50       4.95     12680.0          France    14.85  

[541909 rows x 9 columns]

# Task 9: For each transaction (`InvoiceNo`), calculate the final order amount. Specify the top 5 transactions (`InvoiceNo`) by order amount as the answer.

In [12]:
# Group data by transaction identifier ('InvoiceNo') 
# Sum values in the 'Revenue' column for each transaction (final order amount)  
# Sort transactions by final order amount ('Revenue') in descending order
# Select the top 5 transactions with the highest order amounts 
retail.groupby('InvoiceNo', as_index=False) \
    .agg({'Revenue': 'sum'}) \
    .sort_values('Revenue', ascending=False).head(5)

InvoiceNo    Revenue
22025    581483  168469.60
2303     541431   77183.60
18776    574941   52940.94
19468    576365   50653.91
9741     556444   38970.00